src/firefly/js/externalSource/__tests__/sprintf-test.js

```diff
@@ -0,0 +1,98 @@
+
+        expect(sprintf('%2$s %3$s a %1$s', 'cracker', 'Polly', 'wants')).toBe('Polly wants a cracker');
+        expect(sprintf('Hello %(who)s!', {who: 'world'})).toBe('Hello world!');
+
+        expect(sprintf('%t', true))     .toBe('true');
+        expect(sprintf('%.1t', true))   .toBe('t');
+        expect(sprintf('%t', 'true'))   .toBe('true');
+        expect(sprintf('%t', 1))        .toBe('true');
+        expect(sprintf('%t', false))    .toBe('false');
+        expect(sprintf('%.1t', false))  .toBe('f');
+        expect(sprintf('%t', ''))       .toBe('false');
+        expect(sprintf('%t', 0))		 .toBe('false');
+
+        expect(sprintf('%T', undefined))    .toBe('undefined');
+        expect(sprintf('%T', null))		 .toBe('null');
+        expect(sprintf('%T', true))		 .toBe('boolean');
+        expect(sprintf('%T', 42))           .toBe('number');
+        expect(sprintf('%T', 'This is a string')).toBe('string');
+        expect(sprintf('%T', Math.log))     .toBe('function');
+        expect(sprintf('%T', [1, 2, 3]))    .toBe('array');
+        expect(sprintf('%T', {foo: 'bar'})) .toBe('object');
+        expect(sprintf('%T', /<('[^']*'|'[^']*'|[^''>])*>/)).toBe('regexp');
+
+        expect(sprintf('%v', true))         .toBe('true');
+        expect(sprintf('%v', 42))           .toBe('42');
+        expect(sprintf('%v', [1, 2, 3]))    .toBe('1,2,3');
+        expect(sprintf('%v', 'This is a string'))   .toBe('This is a string');
+        expect(sprintf('%v', {foo: 'bar'}))         .toBe('[object Object]');
+        expect(sprintf('%v', /<("[^"]*"|'[^']*'|[^'">])*>/)).toBe('/<("[^"]*"|\'[^\']*\'|[^\'">])*>/');
+    });
+
+    test('g format:', () => {
+        // If value is greater than or equal to 10^-4 but less than 10^precision then it is represented in decimal format, otherwise scientific
+        //precision is default to 6 if not given.
+        expect(sprintf('%g', 0))        .toBe('0.00000');     // 0
+        expect(sprintf('%g', 123456))   .toBe('123456');      // less than 10^6 (default)
+        expect(sprintf('%g', 123456))   .toBe('123456');      // less than 10^6 (default)
+        expect(sprintf('%g', 1.234567)) .toBe('1.23457');
+        expect(sprintf('%.4g', 123456)) .toBe('1.235e+5');    // less than 10^4.. p=4
+
+        expect(sprintf('%g', 0.0000123456))     .toBe('1.23456e-5');  // less than 10^-4, p=default
+        expect(sprintf('%.4g', 0.0000123456))   .toBe('1.235e-5');    // less than 10^-4, p=4
+    });
+
+    test('J format:', () => {
+        // Java toString() format:   https://docs.oracle.com/javase/7/docs/api/java/lang/Float.html#toString(float)
+        expect(sprintf('%J', 0))            .toBe('0.0');
+        expect(sprintf('%J', 0.123456))     .toBe('0.123456');
+        expect(sprintf('%J', 0.000123456))  .toBe('1.23456e-4');
+        expect(sprintf('%J', 1234567))      .toBe('1234567');
+        expect(sprintf('%J', 12345678))     .toBe('1.2345678e+7');
+    });
```

Comment:
So this one is non-standard format for our convenience? Our default format?

Reply:
This is a new format I added.  It mimic the behavior of Java's toString() on numeric values.  And yes, this is the default format on the client side because we want to match toString() on the server.

Reply:
this is really good idea.